### PR TITLE
Signup: template-first theme Signup flow

### PR DESCRIPTION
**Pull request by michaeldcain**
This adds a Signup flow that allows a user to select a template-first theme instead of having one auto-assigned via segment and topic during Signup.

Epic: https://github.com/Automattic/zelda-private/issues/119

Screen Shot 2019-10-21 at 2 15 29 PM

**To Test:**
- navigate to `/start/design-first/`
- you should see a screen with 18 template-first themes
- select a theme
- finish Signup and verify that the correct theme is set with content similar to the screenshot
- navigate to `/start/domain/`
- create a new site, when you get to the themes step, you should only see 3 themes (based on the selection from the previous step)
- finish Signup and verify that the correct theme is set
- navigate to `/start/domain` and create a domain only site
- click the **Create Site** button on the domain management screen
- go through the `site-selected` flow, picking one of the 3 themes
- Finish flow. Site should be created, but it's not headstarted with the content from the theme (may or may not be a bug, but it's what happens on production)

### CODE REVIEW

File: client/signup/step-wrapper/index.jsx [hunk 1]
```diff
@@ -71,7 +71,7 @@ class StepWrapper extends Component {
 						flowName={ this.props.flowName }
 						stepName={ this.props.stepName }
 						labelText={ this.props.skipLabelText }
-						cssClass={ !! this.props.skipHeadingText && ' navigation-link--has-skip-heading ' }
+						cssClass={ ! this.props.skipHeadingText && ' navigation-link--has-skip-heading ' }
```
Comment (p-jackson) [inline client/signup/step-wrapper/index.jsx]: This change doesn't seem to have an effect on the theme picking steps, but the logic doesn't seem right here. I feel like `has-skip-heading` would be applied if the heading text was present/non-empty.

But yeah, I can't see what effect this has on our flows. I know jetpack use the skip button more though.

File: client/lib/signup/step-actions.js [hunk 2]
```diff
@@ -169,9 +169,10 @@ export function createSiteWithCart( callback, dependencies, stepData, reduxStore
 	const flowToCheck = flowName || lastKnownFlow;
 
 	if (
-		flowToCheck === 'onboarding' &&
-		newSiteParams.options.theme !== defaultSiteTypeTheme &&
-		abtest( 'verticalSuggestedThemes' ) === 'test'
+		get( getSignupDependencyStore( state ), 'useThemeHeadstart' ) ||
```
Comment (michaeldcain) [inline client/lib/signup/step-actions.js]: If the theme step is present in the flow, we should default to theme annotations.

File: client/signup/step-wrapper/index.jsx [hunk 3]
```diff
@@ -71,7 +71,7 @@ class StepWrapper extends Component {
 						flowName={ this.props.flowName }
 						stepName={ this.props.stepName }
 						labelText={ this.props.skipLabelText }
-						cssClass={ !! this.props.skipHeadingText && ' navigation-link--has-skip-heading ' }
+						cssClass={ this.props.skipHeadingText && 'navigation-link--has-skip-heading' }
```
Comment (michaeldcain) [inline client/signup/step-wrapper/index.jsx]: Unrelated change that was causing propType warnings on the first step when `skipHeadingText` was being cast to a bool of `false`.

File: client/signup/steps/theme-selection/index.jsx [hunk 4]
```diff
@@ -23,19 +23,23 @@ import { isEnabled } from 'config';
...
+import './style.scss';
+
 class ThemeSelectionStep extends Component {
 	static propTypes = {
 		designType: PropTypes.string,
+		quantity: PropTypes.number,
 		goToNextStep: PropTypes.func.isRequired,
 		signupDependencies: PropTypes.object.isRequired,
 		stepName: PropTypes.string.isRequired,
-		translate: PropTypes.func,
```
[the remark this answers is not in the document]
Comment (michaeldcain) [reply client/signup/steps/theme-selection/index.jsx]: I wish there were logic for my decisions. I think I removed this as part of some propType investigation. Good catch.

File: client/signup/config/flows-pure.js [hunk 5]
```diff
@@ -99,6 +99,13 @@ export function generateFlows( {
 			lastModified: '2019-08-20',
 		},
 
+		'design-first': {
+			steps: [ 'template-first-themes', 'user', 'domains', 'plans' ],
```
[the remark this answers is not in the document]
Comment (michaeldcain) [reply client/signup/config/flows-pure.js]: Good catch - the flow was originally just a PoC to see if the `themes` step would work, and grew from there. I've added the steps.

File: client/signup/step-wrapper/index.jsx (hunk 1 above)
Comment (michaeldcain) [reply client/signup/step-wrapper/index.jsx]: https://github.com/Automattic/wp-calypso/pull/36825#discussion_r343641893

**Comment by michaeldcain** [comment]
> I like it! I'm wondering whether we want to hide the skip button on the template-first-themes step, it doesn't seem in the spririt of "design first" to not pick a design.

I think it's ok to leave. I don't want the limited number of themes to be a blocker. @ianstewart @jancavan: do y'all have thoughts?

**Comment by ianstewart** [comment]
> I think it's ok to leave. I don't want the limited number of themes to be a blocker.

+1 for the same reasons. What will be the default option here if I skip?

**Comment by ianstewart** [comment]
Will we have tracks events and/or props for theme selected and "skip"?

**Comment by michaeldcain** [comment]
> What will be the default option here if I skip?

Maywood?

**Comment by jancavan** [comment]
> I think it's ok to leave. I don't want the limited number of themes to be a blocker.

Yes, we should keep the Skip link. Some users may not want/be ready to pick a design/not find any of the designs compelling.

**Comment by ianstewart** [comment]
> What will be the default option here if I skip?

Would it be Maywood with the default starter content for the segment? (Blogs get a blog, businesses get a one-page Maywood site.)

**Comment by michaeldcain** [comment]
> Will we have tracks events and/or props for theme selected and "skip"?

- `calypso_signup_theme_select` with prop `theme`
- `calypso_signup_skip_step` with prop `step: template-first-themes`

> Would it be Maywood with the default starter content for the segment? (Blogs get a blog, businesses get a one-page Maywood site.)

@ianstewart: I just pushed a change to make this the case.

**Comment by p-jackson** [comment]
Rebased to hopefully fix failing tests #37400

**Comment by ianstewart** [comment]
@michaeldcain was just testing this flow again not as new-user, but as logged-in user, starting directly from the /design-start screen. Content didn't fully headstart. There was no checklist item for udating my homepage.

**Comment by ianstewart** [comment]
Can we re-order some of the steps in this test to better mirror some of the other irons in the fire right now? So that we'd have …

- Segment step
- Topic step
- Theme step

(I also think it'll be easier to segment and compare with some of the other more recent tests that we've done if we do this.)

**Comment by p-jackson** [comment]
Looks like the e2e failures are explained by p1573732653099600-slack-C1A1EKDGQ
Only an issue when store is sandboxed.

**Comment by ianstewart** [comment]
cc @apeatling

**Comment by michaeldcain** [comment]
I fixed up the theme issue, but am going to go ahead and leave the step where it is, since we're no longer going to be running the AB test with this (although the step will be used in the `test-fse` flow eventually - See #37870).

Should be ready for a re-review.

**Comment by apeatling** [comment]
Can we remove the slow scroll to top after picking the theme? Set the scrolltop to zero without animation when hitting the next step? The scrolling makes the transition feel slow and clunky.

**Comment by apeatling** [comment]
Nice work on this so far. A few questions:

* We have themes in this step that will not be FSE compatible at the end of the milestone, will these be removed? I think it might be okay to include some non-fse blogging themes to bridge the gap.
* Can we order themes based on the data we have for theme launch rate?
* I'm assuming this will be moved to the step position in the figma file?

**Comment by michaeldcain** [comment]
@apeatling: I'll look into the scroll speed. It's noticeably slower b/c of the length of the page with all the options.

> We have themes in this step that will not be FSE compatible at the end of the milestone, will these be removed?

This has become more just a PR to ship changes to the design-first step to update the design and make it work with other onboarding steps. The themes can be updated by just adding or removing them from [the template-first list](https://github.com/Automattic/wp-calypso/pull/36825/files#diff-3d834c6ccb67f206fbcbd4616afa9c3f).

> I think it might be okay to include some non-fse blogging themes to bridge the gap.
Can we order themes based on the data we have for theme launch rate?

Themes are currently randomized. I'd argue that we'd get clearer user intent data this way (i.e. are users just picking the first theme in the list?).

> I'm assuming this will be moved to the step position in the figma file?

Yep. I'll do that as part of a PR to directly address #37870.

**Comment by p-jackson** [comment]
@michaeldcain I've changed the scroll-to-top animation so it takes a constant amount of time regardless of the length of the page.
Still, it can be a little jarring seeing all those images wizz past, especially without another animation for transitioning to the next step.

Perhaps it'd be better to add a `skipScrollAnimation` parameter to `goToStep()` so this particular step can opt-out.

**Comment by p-jackson** [comment]
Looks like this has broken the `/start/domain` flow. Working on fix now.

**Pull request merged**